### PR TITLE
feat: allow specifying whether you want sdists or not

### PR DESCRIPTION
**Pull request by baszalmstra**
This adds the option to specify whether you want to use sdists, prefer them or only use sdists.

### CODE REVIEW

**Comment by notatallshaw** [comment]
> you want to use sdists, prefer them or only use sdists

Sorry, I've not quite got the hand of reading Rust PRs yet, but please make sure there's an option for "prefer not to use them",and IMO that should be the default option), if that in fact is not already the case here.

**Comment by baszalmstra** [comment]
This PR adds the following options:

* Normal (the default): this means we consider any version that has wheels and/or sdists and we sort them based on version.
* PreferWheels: this means we still consider both sdists and wheels but only if no version could be found that has wheels we consider versions with an sdist.
* PreferSdists: this is the opposite of the above. We consider versions with both artifacts but select a version with a sdist over that with a wheel.
* OnlyWheels: dont take sdists into consideration at all, only wheels. No solution is found if there are no wheels. 
* OnlySDists: dont take wheels into consideration at all, only sdists. No solution is found if there are no sdists.

Do you think that suffices?

**Comment by notatallshaw** [comment]
Options seem great, but the default comes with a lot of nuance and introduces backwards compatability issues if you need to change much later down the road.

I would _personally_ suggest `PreferWheels`, a wheel is less likely to have issues than an sdist, and this option is only going to go wrong if a project goes from publishing wheels to only publishing sdists, which would be highly unusual.

There's bewen a discussion in Pip world about changing the default and the problem with backwards compatability for them: https://github.com/pypa/pip/issues/9140 (sorry long discussion)

File: crates/rattler_installs_packages/src/resolve.rs [hunk 1]
```diff
@@ -499,6 +566,51 @@ pub struct PinnedPackage<'db> {
     pub artifacts: Vec<&'db ArtifactInfo>,
 }
 
+/// Defines how to handle sdists during resolution.
+#[derive(Default, Clone, Copy, Eq, PartialOrd, PartialEq)]
+pub enum SDistResolution {
+    /// Treat sdists the same as wheels.
```
Comment (tdejager) [inline crates/rattler_installs_packages/src/resolve.rs:572]: I think this should be worded a bit differently, in the same version `wheels` are preffered but an sdist would be chosen over an older `wheel` version.

**Review by tdejager — DISMISSED** [review]

**Comment by baszalmstra** [comment]
@tdejager I added thorough docs to the `SDistResolution` struct. WDYT?

**Comment by baszalmstra** [comment]
@pradyunsg what do you think of these options? What do you think should be the default here?

**Comment by notatallshaw** [comment]
> Like here 😢 : https://pypi.org/simple/pysdl2/

Oh wow, 🙁, well Normal is definetly universally safe for rip. But the common example of user pain is when there's a new release and wheels are not released along side sdists immediatly.

**Review by wolfv — APPROVED** [review]

**Comment by pradyunsg** [comment]
> what do you think of these options?

They're reasonable, although given that you have low backwards compatibility costs; I suggest using more granular control (eg: enabling people to say "get me X, Y&Z from an sdist with everything else being wheels".

> What do you think should be the default here?

Matching pip's default semantic of "use a wheel if available for a version, otherwise sdist for that version" is likely going to be the most compatible with existing user workflows.

**Pull request merged**